cmd/thanos/rule.go

```diff
@@ -256,10 +268,56 @@ func runRule(
 	reg.MustRegister(rulesLoaded)
 	reg.MustRegister(ruleEvalWarnings)
 
-	for _, addr := range queryAddrs {
-		if addr == "" {
-			return errors.New("static querier address cannot be empty")
+	var queryCfg []query.Config
+	if len(queryConfigYAML) > 0 {
+		var err error
+		queryCfg, err = query.LoadConfigs(queryConfigYAML)
+		if err != nil {
+			return err
+		}
+	} else {
+		for _, addr := range queryAddrs {
+			if addr == "" {
+				return errors.New("static querier address cannot be empty")
+			}
+		}
+
+		// Build the query configuration from the legacy query flags.
+		var fileSDConfigs []http_util.FileSDConfig
+		if len(querySDFiles) > 0 {
+			fileSDConfigs = append(fileSDConfigs, http_util.FileSDConfig{
+				Files:           querySDFiles,
+				RefreshInterval: model.Duration(querySDInterval),
+			})
+		}
+		queryCfg = append(queryCfg,
+			query.Config{
+				EndpointsConfig: http_util.EndpointsConfig{
+					Scheme:          "http",
+					StaticAddresses: queryAddrs,
+					FileSDConfigs:   fileSDConfigs,
+				},
+			},
+		)
+	}
+
+	queryProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix("thanos_ruler_query_apis_", reg),
+		dns.ResolverType(dnsSDResolver),
+	)
+	var queryClients []*http_util.FanoutClient
+	for _, cfg := range queryCfg {
+		c, err := http_util.NewClient(cfg.HTTPClientConfig, "query")
+		if err != nil {
+			return err
 		}
+		c.Transport = tracing.HTTPTripperware(logger, c.Transport)
+		queryClient, err := http_util.NewFanoutClient(logger, cfg.EndpointsConfig, c, queryProvider.Clone())
```

Comment:
Fanout? I think the logic for rule evaluation might be different. It's fanout = broadcast approach.

Reply:
Yea, the logic is ok, just the Name of client might be confusing. It's not Fanout for Querier.